### PR TITLE
add whereNot() to QueryBuilder

### PR DESCRIPTION
**Pull request by adamhigerd**
### Intent

I needed the ability to express `NOT` in a where clause using the query builders.

### Approach

I decided that the most flexible way to do it without causing a combinatorial explosion of function variants was to wrap a compound where clause with a `whereNot()` method.

### Automated Tests

Added a unit test for it.

### QA Notes

> Ensure you have updated the QA Notes in the original issue.

### Documentation

API docs are in the code.

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md`
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

**Review by kevinushey — COMMENTED** [review]
Nice, this is a useful addition. Two behaviors worth calling out (and ideally pinning down with tests / a doc note), since both only show up on the delete / update paths where a wrong row set is hard to notice.

### 1. Negating an empty compound clause matches every row

The existing convention is that an empty `QBWhereAnd()` renders as `(1 = 0)` ("no conditions => match nothing"). With `negate`, that flips to `(1 = 1)`. That's internally consistent, but it's the opposite of what ordinary logic suggests (an empty AND is vacuously true, so its negation should match nothing), and it's easy to hit by accident via `whereIfSet()`:

```cpp
// "delete sessions, except the protected ones"
DeleteBuilder del(conn, "active_session_metadata");
del.whereNot(QBWhereAnd()
   .whereIfSet("user_id", keepUser)
   .whereIfSet("workbench", keepWorkbench));
```

| inputs | generated SQL | effect |
|---|---|---|
| both set | `DELETE FROM active_session_metadata WHERE NOT (user_id = :... AND workbench = :...)` | as intended |
| neither set | `DELETE FROM active_session_metadata WHERE (1 = 1)` | deletes every row |

The same shape with `QBWhereOr()` ("delete sessions matching none of these labels") goes the other way and silently deletes nothing when no labels are supplied.

I don't think the mapping itself is wrong, but nothing in `DatabaseTests.cpp` covers the empty-compound branches (there's no `1 = 0` / `1 = 1` assertion at all), so a later tweak to the un-negated arm could invert the negated one without anything going red. Suggest adding `toSQL()` assertions for `whereNot(QBWhereAnd())` -> `(1 = 1)` and `whereNot(QBWhereOr())` -> `(1 = 0)`, or computing the un-negated sentinel once and inverting it so the two arms can't drift.

### 2. `NOT (...)` drops rows where a referenced column is NULL

`whereNot()` emits plain `NOT (...)`, so it inherits SQL three-valued logic. The positive path already hides this from callers (`where(column, nullptr)` emits `IS NULL`), so people are used to the builder handling NULL for them. Several columns in `active_session_metadata` are nullable (`project`, `r_version`, `suspend_timestamp`, ...):

```cpp
// "all sessions not open on this project"
select.whereNot(QBWhereAnd().where("project", projectPath));
```

generates

```sql
SELECT ... FROM active_session_metadata WHERE NOT (project = :where_1_project)
```

For a session with no open project, `project = :p` is NULL, `NOT NULL` is still NULL, and both SQLite and PostgreSQL treat that as not matched. So every project-less session is excluded from the "not this project" result. A sweep for "sessions not suspended" against `suspend_timestamp` has the same hole.

A short caveat in the `whereNot()` comment ("rows where a referenced column is NULL are not matched; add an explicit `where(col, nullptr)` in an OR if you need them") would probably be enough for now, since fixing it properly means an `IS NOT NULL` counterpart on the negated path.

### Minor

- The new test only asserts `toSQL()`; the two sibling cases in the same test also `build()` and execute against SQLite. Executing `NOT (bar = 0 AND foo = 2)` over the seeded rows should return ids `{3, 5}`, which would also catch any prefix mismatch between `toClause()` and `apply()` for negated nodes.
- `QueryBuilder.hpp` still says `QBWhereAnd` is "only practically useful within a `QBWhereOr()` expression", which this PR makes untrue; worth updating alongside a `whereNot()` example there.
- `ss << "NOT ";` in `QBNodeList.hpp` is indented 8 spaces rather than 9.

File: src/cpp/core/include/core/details/QBNodeList.hpp [hunk 1]
```diff
@@ -104,15 +104,17 @@ struct QBCompoundNode : public QBBaseNode
       // WHERE () is invalid SQL
       if (nodes.empty())
       {
-         // AND implies that no records should be matched if no conditions match
-         if (op == "AND")
+         // AND / NOT OR implies that no records should be matched if no conditions match
+         if (negate ? (op == "OR") : (op == "AND"))
```
Comment (gtritchie) [inline src/cpp/core/include/core/details/QBNodeList.hpp:118]: 🟠 Negating an empty compound clause turns the "match nothing" fallback into `WHERE (1 = 1)`, so an UPDATE or DELETE hits every row.

Consider returning `(1 = 0)` whenever the clause has no conditions, negated or not, so an empty clause can never widen a query.

<details><summary>Why</summary>

`DeleteBuilder d(conn, "jobs"); d.whereNot(QBWhereAnd().whereIfSet("id", boost::optional<int>()));` — the optional is unset, so the compound node ends up with no children. `toClause()` takes the `nodes.empty()` branch; `negate` is true and `op` is `"AND"`, so the ternary is false and it falls through to `(1 = 1)`. The generated SQL is `DELETE FROM jobs WHERE (1 = 1)`.

The inversion is logically correct — `NOT false` is `true` — but the existing `(1 = 0)` isn't really logic, it's a guard: an empty AND is logically TRUE, and the code returns FALSE anyway so that a clause that collapsed to nothing matches nothing. Inverting the guard gives back exactly the behavior it was there to prevent.

Not merge-blocking because `whereNot()` is new and has no callers yet, so nothing in the tree can reach it today.

</details>

File: src/cpp/core/DatabaseTests.cpp [hunk 2]
```diff
@@ -1232,6 +1232,16 @@ TEST_F(DatabaseTestsFixture, CompoundWhereExpressions)
...
    ASSERT_FALSE(unexpected) << "unexpected results in WHERE AND test";
+
+   SelectBuilder builder3(sqliteConnection, "BuilderTest");
+   builder3.add("id");
+   builder3.whereNot(QBWhereAnd()
+      .where("bar", 0)
+      .where("foo", 2)
+   );
+
+   sql = builder3.toSQL();
```
Comment (gtritchie) [inline src/cpp/core/DatabaseTests.cpp:1243]: 🟠 This case stops at the generated SQL string, so parameter binding and actual row selection for a negated clause are never exercised.

Follow the two cases above it: call `build()`, `execute()`, and assert the returned ids.

<details><summary>Why</summary>

If `QBCompoundNode::apply()` bound parameters under the wrong prefix for a negated node, or a backend rejected the generated `NOT (...)` form, this test would still pass — it never gets as far as running the query. `builder` and `builder2` in the same `TEST_F` both execute and check the resulting ids; `builder3` doesn't.

</details>

File: src/cpp/core/include/core/details/QBMixins.hpp [hunk 3]
```diff
@@ -266,6 +266,15 @@ class QBWhereMixin : virtual public QueryBuilder
       return std::move(*self);
    }
 
+   // As where(), but negates the match
+   BUILDER&& whereNot(QBCompoundWhere&& clause)
```
Comment (gtritchie) [inline src/cpp/core/include/core/details/QBMixins.hpp:270]: 🟡 "negates the match" isn't the whole story: `NOT (col = :v)` is unknown when `col` is NULL, so a NULL row matches neither `where()` nor `whereNot()`.

Mention the NULL behavior in the comment.

<details><summary>Why</summary>

Given a row with `foo` NULL, `where(QBWhereAnd().where("foo", 2))` correctly omits it, and `whereNot(QBWhereAnd().where("foo", 2))` omits it too. A caller reading the comment would expect the two queries to partition the table and will silently lose those rows from both result sets.

</details>

File: src/cpp/core/include/core/details/QBNodeList.hpp [hunk 4]
```diff
@@ -104,15 +104,17 @@ struct QBCompoundNode : public QBBaseNode
...
             return "(1 = 0)";
-         // OR implies that all records should be matched if no conditions fail
+         // OR / NOT AND implies that all records should be matched if no conditions fail
          return "(1 = 1)";
       }
       std::string extPrefix = prefix + refName + "_";
       std::ostringstream ss;
       bool first = true;
+      if (negate)
+        ss << "NOT ";
```
Comment (gtritchie) [inline src/cpp/core/include/core/details/QBNodeList.hpp]: 🟡 Indented two spaces; the rest of the file and `src/cpp/.clang-format` use three.

```suggestion
         ss << "NOT ";
```

**Review by gtritchie — COMMENTED** [review]
**No blocking issues.** 2 follow-up, 2 nits.

🟠 is real but not merge-blocking — either address it here or open an issue to do so later. 🟡 is optional.

- 🟠 `src/cpp/core/include/core/details/QBNodeList.hpp:108` — negated empty clause becomes `WHERE (1 = 1)`
- 🟠 `src/cpp/core/DatabaseTests.cpp:1243` — new test asserts SQL text only, never runs the query
- 🟡 `src/cpp/core/include/core/details/QBMixins.hpp:270` — doc comment omits NULL behavior
- 🟡 `src/cpp/core/include/core/details/QBNodeList.hpp:117` — indentation

**Comment by gtritchie** [comment]
Double review! I bet you feel special now.

File: src/cpp/core/include/core/details/QBNodeList.hpp (hunk 4 above)
Comment (adamhigerd) [reply src/cpp/core/include/core/details/QBNodeList.hpp]: Huh. Whoops. That's odd, my editor usually catches that one too. Thanks.

File: src/cpp/core/include/core/details/QBMixins.hpp (hunk 3 above)
Comment (adamhigerd) [reply src/cpp/core/include/core/details/QBMixins.hpp:270]: Hmm... Nitpicky, but valid. Okay, I'll update it.

File: src/cpp/core/DatabaseTests.cpp (hunk 2 above)
Comment (adamhigerd) [reply src/cpp/core/DatabaseTests.cpp:1243]: The LLMs tend to be pretty excessive in their unit testing habits. The test that I added covers the new code path. QBCompoundWhere is already covered sufficiently well.

File: src/cpp/core/include/core/details/QBNodeList.hpp (hunk 1 above)
Comment (adamhigerd) [reply src/cpp/core/include/core/details/QBNodeList.hpp:118]: I disagree now just as much as I did last time the LLM complained about this construction. The idea here is that you could set up a `QBWhereAnd` and dynamically add conditions to exclude rows, or a `QBWhereOr` and dynamically add conditions to include them. If you add no exclusions, then you include everything. If you add no inclusions, you exclude everything. Ideally you just don't write that kind of code in the first place, but at least it's documented and consistent.

**Comment by adamhigerd** [comment]
> but nothing in DatabaseTests.cpp covers the empty-compound branches (there's no 1 = 0 / 1 = 1 assertion at all), so a later tweak to the un-negated arm could invert the negated one without anything going red

Hm. I suppose now that I've doubled down on the intent behind the current behavior, it's no longer really an edge case that could be modified without practical impact. Valid point, there should be a test case.

> 2. NOT (...) drops rows where a referenced column is NULL

I'll update the comment to be more precise.

> The new test only asserts toSQL(); the two sibling cases in the same test also build() and execute against SQLite. Executing NOT (bar = 0 AND foo = 2) over the seeded rows should return ids {3, 5}, which would also catch any prefix mismatch between toClause() and apply() for negated nodes.

As I also stated in response to Gary's LLM, there's no need to add the redundant tests when QBCompoundWhere already has sufficient coverage.

> QueryBuilder.hpp still says QBWhereAnd is "only practically useful within a QBWhereOr() expression", which this PR makes untrue; worth updating alongside a whereNot() example there.

Hm. Valid point. I'll update that.

> ss << "NOT "; in QBNodeList.hpp is indented 8 spaces rather than 9.

Will fix.

File: src/cpp/core/include/core/details/QBNodeList.hpp (hunk 1 above)
Comment (adamhigerd) [reply src/cpp/core/include/core/details/QBNodeList.hpp:118]: ... wait a sec... *headscratch* Heh, okay, so I stand by my rationale, but my implementation was wrong the whole time!

**Comment by adamhigerd** [comment]
Okay, I think that should cover everything.

File: src/cpp/core/include/core/details/QBNodeList.hpp [hunk 5]
```diff
@@ -101,18 +101,30 @@ struct QBCompoundNode : public QBBaseNode
...
+         // An OR clause includes records that match any condition.
+         // Therefore, an empty OR clause INCLUDES no records.
+
+         // A NOT clause (interactions with NULL notwithstanding) includes
+         // records that would have been excluded, and vice versa.
+         // Therefore, an empty NOT(AND) clause INCLUDES no records,
+         // and an empty NOT(OR) clause EXCLUDES no records.
+
+         if (negate ? (op == "OR") : (op == "AND"))
+            return "(1 = 1)";
```
Comment (gtritchie) [inline src/cpp/core/include/core/details/QBNodeList.hpp:119]: 🟠 This also flips the un-negated cases: an empty `where(QBWhereAnd())` on an update or delete used to touch no rows and now touches all of them.

Check existing callers for compound clauses that can end up empty, and call out the semantic change in the PR description.

<details><summary>Why</summary>

On `main`, empty `QBWhereAnd` renders `(1 = 0)` and empty `QBWhereOr` renders `(1 = 1)`. After this commit they're `(1 = 1)` and `(1 = 0)` respectively, independent of `whereNot()`:

```cpp
DeleteBuilder del(conn, "some_table");
del.where(QBWhereAnd()
   .whereIfSet("user_id", userId)        // unset
   .whereIfSet("workbench", workbench)); // unset
// main:  DELETE FROM some_table WHERE (1 = 0)   -- no rows
// now:   DELETE FROM some_table WHERE (1 = 1)   -- every row
```

The new mapping is the logically correct one and is now documented, so I'm not suggesting reverting it. The concern is existing code written against the old behavior. The only in-tree callers are tests, and none of them builds an empty clause, but QueryBuilder was backported (#18433), so there may be callers outside this repo that relied on it. The commit message ("applying PR review feedback") and the PR description don't mention the change.

</details>

**Review by gtritchie — APPROVED** [review]
LGTM, one more finding posted but up to you if it's worth addressing.

Recommend merging `main` into the branch so CI re-runs (we're making lots of e2e test hardening fixes to try and make them less flakey).

File: src/cpp/core/include/core/details/QBNodeList.hpp (hunk 5 above)
Comment (adamhigerd) [reply src/cpp/core/include/core/details/QBNodeList.hpp:119]: There is no existing code and there are now unit tests to validate the behavior.
